### ISSUE TITLE
Inline pep484 typing in the SQLAlchemy source

### ISSUE THREAD

**Issue by CaselIT** [sqlalchemy/sqlalchemy#6810]
The plan would be to phase out the stubs package at https://github.com/sqlalchemy/sqlalchemy2-stubs and move all the type annotation inline with the source.

Targeting for 2.0 but this may realistically be pushed to 2.x

**Comment by CaselIT** [comment]
We should try https://github.com/ambv/retype/

**Comment by sqla-tester** [comment]
**Federico Caselli** referenced this issue:

**Properly type _generative** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3423

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **main** branch:

**initial reorganize for static typing** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3447

**Comment by zzzeek** [comment]
related issues:

#7535 
#7519

**Comment by zzzeek** [comment]
also there's the issue of replacing existing sqlalchemy-stubs packages in type checkers and it's suggested we use a "positive" masking approach, as a "negative" uninstall approach is not supported by type checkers that bundle the stubs: https://github.com/microsoft/pylance-release/issues/840#issuecomment-1008329878

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **main** branch:

**WIP for ORM typing** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3495

**Comment by zzzeek** [comment]
current goals are:

1. get https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3495, based on https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3447, working.   POC here is that I would like to include updating the mypy plugin and removing the dependency on sqlalchemy2-stubs, w/ mypy tests passing.  that will be the proof of concept for 3495, which will also validate 3447.
2. then merge 3447 and 3495
3. after those two are in, that's the "hard part" of typing, that is, the column expression / operator / ORM mapping stuff.  then we can do a package-by-package or module-by-module approach in separate gerrits 
4. separately, when doing the typing I hit a lot of blocks which are due to the fact that we use metaclasses for "visitable" items as well as some other things.  I want to see how far we can use `__subclass_init__` to get rid of all metaclasses, at the very least the metaclasses used internally by "visitable" as well as functions.py etc.   The DeclarativeMeta metaclass might be a different situation if this is going to break backwards compatibility.

**Comment by CaselIT** [comment]
Step 4. was something that I meant to suggest since we now are guaranteed to always have at least py3.6

does it make sense as a separate change?

**Comment by zzzeek** [comment]
im going to try it now as its own thing, not including declarative, which probably also needs to be its own thing

**Comment by zzzeek** [comment]
I think we might want to skip declarative as the DeclarativeMeta class also intercepts class-level setattr and delattr events.  in that sense we are still using metaclass-only features.  

i have all the internal stuff off of metaclasses with one small exception in the ORM, will keep it at that.

**Comment by sqla-tester** [comment]
**Mike Bayer** referenced this issue:

**remove internal use of metaclasses** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3499

**Comment by sqla-tester** [comment]
**Mike Bayer** referenced this issue:

**Initial ORM typing layout** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3495

**Comment by CaselIT** [comment]
Re opened since there is still lot to do

**Comment by sqla-tester** [comment]
**Mike Bayer** referenced this issue:

**mypy: sqlalchemy.util** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3525

**Comment by sqla-tester** [comment]
**Mike Bayer** referenced this issue:

**get mypy test suite to work, take 2** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3562

**Comment by sqla-tester** [comment]
**Mike Bayer** referenced this issue:

**pep-484 inline typing for sqlalchemy.event** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3590

**Comment by zzzeek** [comment]
ensure assocaition proxy collection implement typing.List/Set/Dict, see #7774

**Comment by sqla-tester** [comment]
**Mike Bayer** referenced this issue:

**pep484 + abc bases for assocaitionproxy** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3637

**Comment by sqla-tester** [comment]
**Mike Bayer** referenced this issue:

**pep-484: sql** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3662

**Comment by sqla-tester** [comment]
**Mike Bayer** referenced this issue:

**pep484 for hybrid** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3710

**Comment by sqla-tester** [comment]
**Mike Bayer** referenced this issue:

**pep484 - sql.selectable** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3758

**Comment by zzzeek** [comment]
I'm going to merge https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3825 and that will be it for now, where of course we can keep typing things, but as of now, the majority of SQLAlchemy's public-facing APIs are typed, as well as core internals that are dependencies for other things.  Internals that are further out in the dependency sphere (namely dialects as well as lesser used ORM extensions, ORM query and persistence internals) are largely un-typed.

Will sum up what we have:

## Completely Typed - passes mypy strict ##

* sqlalchemy.pool
* sqlalchemy.connectors
* sqlalchemy.engine excluding cursor, default, reflection
* sqlalchemy.event
* sqlalchemy.exc, sqlalchemy.log, sqlalchemy.inspection, front-facing sqlalchemy.types , sqlalchemy.schema
* sqlalchemy.sql strict typed modules:
    * annotation
    * cache_key
    * default_comparator
    * _elements_constructors
    * dml
    * _dml_constructors
    * events
    * expression
    * operators
    * roles
    * schema
    * _selectable_constructors
    * type_api
    * visitors
* sqlalchemy.orm strict typed modules:
    * base
    * clsregistry
    * decl_base
    * decl_api
    * descriptor_props
    * exc
    * identity
    * interfaces
    * _orm_constructors
    * path_registry
    * properties
    * query
    * relationships
    * scoping
    * session
    * state
    * state_changes
* sqlalchemy.ext strict typed modules: I got the big ones here:
    * association_proxy
    * asyncio
    * hybrid
    * mypy (was mostly done already)
* sqlalchemy.util.topological


## non-strict typed: passes typing but omits non-typed functions, calls to non-typed functions ##

* sqlalchemy.util (all files excluding topological which is strict typed)
* sqlalchemy.engine, two files:
    * cursor
    * default
* sqlalchemy.sql:
    * base
    * coercions
    * compiler
    * crud
    * ddl
    * elements
    * functions
    * lambdas
    * naming
    * traversals
    * selectable
    * sqltypes
    * util
* sqlalchemy.orm:
     * attributes
     * collections
     * instrumentation
     * mapper
     * strategy_options
     * sync
     * util

## untyped, mypy is set to ignore_errors ##

these are files that have not been passed over. 

* sqlalchemy.orm
    * context
    * dependency
    * dynamic
    * evaluator
    * events
    * mapped_collection
    * persistence
    * loading
    * strategies
    * unitofwork

* sqlalchemy.dialects - no dialect files are run under mypy
* sqlalchemy.testing

* sqlalchemy.ext
    * automap
    * baked
    * declarative.extensions
    * horizontal_shard
    * mutable
    * instrumentation
    * indexable
    * orderinglist
    * serializer


## priority to do ##

Of the above lists, this is where the need for continued typing efforts are most critical

* sqlalchemy.engine.reflection # should be strict, waiting on refactor to be merged

* sqlalchemy.sql.sqltypes  # should be strict
* sqlalchemy.sql.functions  # should be strict
* sqlalchemy.sql.selectable  # everything public should be typed
* sqlalchemy.sql.elements  # everything public should be typed

* sqlalchemy.orm.dynamic  # should be partially typed
* sqlalchemy.orm.events  # should be strict
* sqlalchemy.orm.mapped_collection  # should be strict

* sqlalchemy.ext.automap  # should be strict
* sqlalchemy.ext.mutable  # should be strict
* sqlalchemy.ext.horizontal_shard  # should be partially typed, there may still be some overly legacy things here


## important ORM internals that should pass non-strict ##

I spent months on typing and getting these last ORM internals seems like I need to come back to them later, as they will be pretty tedious, but they should have non-strict typing which at the very least is great for finding bugs:

* sqlalchemy.orm.loading
* sqlalchemy.orm.context
* sqlalchemy.orm.persistence
* sqlalchemy.orm.strategies

**Comment by CaselIT** [comment]
sounds good to me mike. do you want to keep this open of close it?

Maybe we can keep open and update the first message with your list?

**Comment by zzzeek** [comment]
probably keep it open and re-target the milestone

**Comment by ZanderBrown** [comment]
Use of `sqlalchemy.orm.validates` triggers

```
error: Untyped decorator makes function "<func>" untyped
```

In practice it's a harmless distraction, buy annoying nevertheless

**Comment by mgmira** [comment]
Postgres apis (such as on_conflict_do_nothing) are still untyped. Is there something blocking it from being fully typed ? Or is this just a matter of manpower to annotate the methods ? :)

**Comment by CaselIT** [comment]
The second mainly

**Comment by sqla-tester** [comment]
**Mike Bayer** referenced this issue:

**fully type functions.py** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4957

**Comment by sqla-tester** [comment]
**Mike Bayer** referenced this issue:

**fully type functions.py** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4958

**Comment by bhperry** [comment]
Mypy raises errors about `MutableDict.as_mutable(JSONB)` after upgrading to sqlalchemy 2.x (with postgresql dialect)
```
Argument 1 to "as_mutable" of "Mutable" has incompatible type "Type[JSONB]"; expected "TypeEngine[Any]"  [arg-type]
``` 


Is this expected in the current state?

**Comment by zzzeek** [comment]
probably, try sending JSONB() instantiated, that will help with typing

**Comment by bhperry** [comment]
That did the trick, thanks!

**Comment by tamird** [comment]
Would it be possible to add annotations for `func.nullif`? I see functions is marked checked, but this is notably missing. I'd be happy to contribute this but need a bit of guidance.

**Comment by tyree731** [comment]
Found an interesting case with the postgres dialect, wherein the order you call `on_conflict_do_update` vs. `returning` matters in a statement, I believe because `returning` returns a generic `ReturningInsert`, whereas `on_conflict_do_update` only works on the postgres-specific `Insert` type. Example:

```
from sqlalchemy.dialects.postgresql import insert
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column


class Base(DeclarativeBase):
    pass


class Foo(Base):
    __tablename__ = "foo"
    __table_args__ = ()
    id: Mapped[int] = mapped_column(primary_key=True)
    bar: Mapped[str]


# example.py:19: error: "ReturningInsert[tuple[int]]" has no attribute "on_conflict_do_update"  [attr-defined]
foo_record = {"id": 1, "bar": "bar"}
statement_1 = insert(Foo).values({"id": 1, "bar": "bar"})
statement_2 = statement_1.returning(Foo.id)
statement_3 = statement_2.on_conflict_do_update(
    index_elements=[Foo.id], set_={"bar": foo_record["bar"]}
)

"""
the below insert code doesn't type error:
statement_1 = insert(Foo).values({"id": 1, "bar": "bar"})
statement_2 = statement_1.on_conflict_do_update(
    index_elements=[Foo.id], set_={"bar": foo_record["bar"]}
)
statement_3 = statement_2.returning(Foo.id)
"""
```

**Comment by kkirsche** [comment]
I was looking at `sqlalchemy.ext.compiler:compiles` today and wanted to share the patch that addressed the typing errors I was running into.

I'm not confident I know what the `element` or `compiler` arguments are to `_dispatcher` class or `_wrap_existing_dispatch` functions, but if those can be defined, would you be interested in this as a pull request?

Thanks for your time and for this great library!

```
diff --git a/lib/sqlalchemy/ext/compiler.py b/lib/sqlalchemy/ext/compiler.py
index 01462ad0b..d9835b1f0 100644
--- a/lib/sqlalchemy/ext/compiler.py
+++ b/lib/sqlalchemy/ext/compiler.py
@@ -452,15 +452,21 @@ Example usage::
     )

 """
diff --git a/lib/sqlalchemy/ext/compiler.py b/lib/sqlalchemy/ext/compiler.py
index 01462ad0b..fe2bc8b52 100644
--- a/lib/sqlalchemy/ext/compiler.py
+++ b/lib/sqlalchemy/ext/compiler.py
@@ -452,15 +452,21 @@ Example usage::
     )
 
 """
+from typing import Any
+from typing import Callable # TODO: Change to collections.abc when 3.8 support is dropped
+from typing import TypeVar
+
 from .. import exc
 from ..sql import sqltypes
 
+_F = TypeVar("_F", bound=Callable[..., Any])
 
-def compiles(class_, *specs):
+
+def compiles(class_: type, *specs: str) -> Callable[[_F], _F]:
     """Register a function as a compiler for a
     given :class:`_expression.ClauseElement` type."""
 
-    def decorate(fn):
+    def decorate(fn: _F) -> _F:
         # get an existing @compiles handler
         existing = class_.__dict__.get("_compiler_dispatcher", None)
 
@@ -505,7 +511,7 @@ def compiles(class_, *specs):
     return decorate
 
 
-def deregister(class_):
+def deregister(class_: type) -> None:
     """Remove all custom compilers associated with a given
     :class:`_expression.ClauseElement` type.
 
@@ -517,7 +523,7 @@ def deregister(class_):
 
 
 class _dispatcher:
-    def __init__(self):
+    def __init__(self) -> None:
         self.specs = {}
 
     def __call__(self, element, compiler, **kw):
```

**Comment by zzzeek** [comment]
hey there -

sure, sending along typing additions where there were none previously is helpful as long as PRs aren't spanning dozens of files, or aren't adding tons of "type: ignores" to make things work (we'd rather get a PR that errors out so we can improve it before adding ignores).

**Comment by sqla-tester** [comment]
**Kevin Kirsche** referenced this issue:

**Add type hints to `sqlalchemy.ext.compiler`** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5512

**Comment by sqla-tester** [comment]
**Kevin Kirsche** referenced this issue:

**Add type hints to `sqlalchemy.ext.compiler`** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5517

**Comment by Polandia94** [comment]
Started to type mysql dialect. I created a drft pull request to document the advancement and have the discusions. If is preferred i can move the discussion to issues or gerrit

**Comment by CaselIT** [comment]
It's fine to talk on the PR. thanks!

**Comment by dlax** [comment]
I started to work on the `postgresql` dialect, first PR https://github.com/sqlalchemy/sqlalchemy/pull/12384 about the `array` module.

**Comment by kapildagur** [comment]
Hi! I'd like to work on adding type annotations to `sqlalchemy.sql.base`.

**Comment by CaselIT** [comment]
sure feel free to propose a pr, but try to keep non typing changes and other module changes to a minimum

**Comment by kapildagur** [comment]
Hi @CaselIT,

When you have a moment, could you please review my [PR for `sqlalchemy.sql.base`](https://github.com/sqlalchemy/sqlalchemy/pull/12707)? Once it's acknowledged, I’d like to plan follow-up typing contributions for another untyped module( [`sqlalchemy.sql.crud`](https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/sql/crud.py) or [`sqlalchemy.sql.compiler`](https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/sql/compiler.py)) as part of issue #6810.

I’m hoping to dedicate some time next weekend to continue this effort, so your feedback or approval would be greatly appreciated.

Thanks!

**Comment by CaselIT** [comment]
Hi,

I'll review that PR, but I can't commit to a timeline at the moment.
Note that the internal modules of sql, while nice to have, are not really critical to have typed, since the api there that's left untyped is almost all private.

**Comment by kapildagur** [comment]
Alright @CaselIT !
I believe that fully typing all `sqlalchemy` modules — without relying on `# type: ignore` — would ultimately lead to more maintainable and robust code.

I’m planning to continue contributing toward that goal, both as part of my learning journey and to support the project.
If the priority is to type public-facing modules first, I’ll shift my focus to the `sqlalchemy.orm` package this weekend. Let me know if you have any preferences or suggestions on which modules might be most useful to tackle next.

**Comment by CaselIT** [comment]
> I believe that fully typing all `sqlalchemy` modules — without relying on `# type: ignore` — would ultimately lead to more maintainable and robust code.

for sure, it's just not the top priority, and that's why progress on it has been a bit slow.

regarding what to do next, the ones you have listed before were fine, but please try to keep the changes as typing only. If you think some code change would make sense, that's ok to propose, but please do them in a separate PR, asking before in an issue or discussion if it's more extensive. 

Thanks for your effort!

**Comment by kapildagur** [comment]
> > I believe that fully typing all `sqlalchemy` modules — without relying on `# type: ignore` — would ultimately lead to more maintainable and robust code.
> 
> for sure, it's just not the top priority, and that's why progress on it has been a bit slow.
> 
> regarding what to do next, the ones you have listed before were fine, but please try to keep the changes as typing only. If you think some code change would make sense, that's ok to propose, but please do them in a separate PR, asking before in an issue or discussion if it's more extensive. 
> 
> Thanks for your effort!

Thanks!!, Your Welcome!!

**Comment by sqla-tester** [comment]
**KapilDagur** referenced this issue:

**typing: improve type coverage in sql.base** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6028

**Comment by sqla-tester** [comment]
**KapilDagur** referenced this issue:

**typing: improve type coverage in sql.base** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6029

**Comment by kapildagur** [comment]
Hi @zzzeek, @CaselIT 

Sorry to interrupt, but I’ve run into a typing-related question while working on the compiler module.

I’ve opened a **temporary PR** to explain the issue and provide context:
**Temp PR:** https://github.com/sqlalchemy/sqlalchemy/pull/12759

Would appreciate it if you could take a look when you get a chance and share your thoughts.

Thanks a lot!

**Comment by kapildagur** [comment]
Hello @CaselIT ,
This weekend I'm planning for `sql.crud` module, Plz ack me so that I'll plan my day...
Thanks!

**Comment by CaselIT** [comment]
ok!

**Comment by sqla-tester** [comment]
**Denis Laxalde** referenced this issue:

**Complete type annotations of sqlalchemy.engine.cursor module** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6187

**Comment by sqla-tester** [comment]
**Denis Laxalde** referenced this issue:

**Complete type annotations of sqlalchemy.engine.cursor module** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6192
